### PR TITLE
Refactor and add validation to IntervalIndex.__init__

### PR DESCRIPTION
**Pull request by mroeschke**
## Description
* Adding validation to `closed`, `dtype` arguments in `ItervalIndex.__init__`
* Ensure `closed` attribute always maps to `IntervalDtype.closed`
* `build_interval_column` was no longer necessary by using `IntervalColumn` directly

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Review by bdice — APPROVED** [review]

File: python/cudf/cudf/core/column/column.py [hunk 1]
```diff
@@ -1697,52 +1697,6 @@ def build_categorical_column(
     return cast("cudf.core.column.CategoricalColumn", result)
 
 
-def build_interval_column(
```
Comment (vyasr) [inline python/cudf/cudf/core/column/column.py:1700]: Amazing! I've been wanting to rip out more of these factories for so long!

**Review by vyasr — APPROVED** [review]

**Comment by bdice** [comment]
@mroeschke Looks like there are some test failures to work out.

```
=========================== short test summary info ============================
FAILED tests/test_series.py::test_series_value_counts_bins[1] - AssertionError: Series.index are different

Series.index values are different (100.0 %)
[left]:  IntervalIndex([(0.997, 3.0]], dtype='interval[float64, right]')
[right]: IntervalIndex([(0.997, 0.997]], dtype='interval[float64, right]')
FAILED tests/test_series.py::test_series_value_counts_bins[2] - AssertionError: Series.index are different

Series.index values are different (100.0 %)
[left]:  IntervalIndex([(0.997, 2.0], (2.0, 3.0]], dtype='interval[float64, right]')
[right]: IntervalIndex([(0.997, 0.997], (2.0, 2.0]], dtype='interval[float64, right]')
FAILED tests/test_series.py::test_series_value_counts_bins[3] - AssertionError: Series.index are different

Series.index values are different (100.0 %)
[left]:  IntervalIndex([(0.997, 1.667], (1.667, 2.333], (2.333, 3.0]], dtype='interval[float64, right]')
[right]: IntervalIndex([(0.997, 0.997], (1.667, 1.667], (2.333, 2.333]], dtype='interval[float64, right]')
FAILED tests/test_series.py::test_series_value_counts_bins_dropna[True-1] - AssertionError: Series.index are different

Series.index values are different (100.0 %)
[left]:  IntervalIndex([(0.997, 3.0]], dtype='interval[float64, right]')
[right]: IntervalIndex([(0.997, 0.997]], dtype='interval[float64, right]')
FAILED tests/test_series.py::test_series_value_counts_bins_dropna[True-2] - AssertionError: Series.index are different

Series.index values are different (100.0 %)
[left]:  IntervalIndex([(0.997, 2.0], (2.0, 3.0]], dtype='interval[float64, right]')
[right]: IntervalIndex([(0.997, 0.997], (2.0, 2.0]], dtype='interval[float64, right]')
FAILED tests/test_series.py::test_series_value_counts_bins_dropna[True-3] - AssertionError: Series.index are different

Series.index values are different (100.0 %)
[left]:  IntervalIndex([(0.997, 1.667], (1.667, 2.333], (2.333, 3.0]], dtype='interval[float64, right]')
[right]: IntervalIndex([(0.997, 0.997], (1.667, 1.667], (2.333, 2.333]], dtype='interval[float64, right]')
FAILED tests/test_series.py::test_series_value_counts_bins_dropna[False-1] - AssertionError: Series.index are different

Series.index values are different (100.0 %)
[left]:  IntervalIndex([(0.997, 3.0]], dtype='interval[float64, right]')
[right]: IntervalIndex([(0.997, 0.997]], dtype='interval[float64, right]')
FAILED tests/test_series.py::test_series_value_counts_bins_dropna[False-2] - AssertionError: Series.index are different

Series.index values are different (100.0 %)
[left]:  IntervalIndex([(0.997, 2.0], (2.0, 3.0]], dtype='interval[float64, right]')
[right]: IntervalIndex([(0.997, 0.997], (2.0, 2.0]], dtype='interval[float64, right]')
FAILED tests/test_series.py::test_series_value_counts_bins_dropna[False-3] - AssertionError: Series.index are different

Series.index values are different (100.0 %)
[left]:  IntervalIndex([(0.997, 1.667], (1.667, 2.333], (2.333, 3.0]], dtype='interval[float64, right]')
[right]: IntervalIndex([(0.997, 0.997], (1.667, 1.667], (2.333, 2.333]], dtype='interval[float64, right]')
FAILED tests/test_indexing.py::test_series_iloc_scalar_interval_return_pd_scalar[iloc-0] - AssertionError: assert Interval(1, 1, closed='right') == Interval(1, 2, closed='right')
FAILED tests/test_indexing.py::test_series_iloc_scalar_interval_return_pd_scalar[loc-a] - AssertionError: assert Interval(1, 1, closed='right') == Interval(1, 2, closed='right')
FAILED tests/test_indexing.py::test_dataframe_iloc_scalar_interval_return_pd_scalar[iloc-0-0] - AssertionError: assert Interval(1, 1, closed='right') == Interval(1, 2, closed='right')
FAILED tests/test_indexing.py::test_dataframe_iloc_scalar_interval_return_pd_scalar[loc-a-a] - AssertionError: assert Interval(1, 1, closed='right') == Interval(1, 2, closed='right')
FAILED tests/test_doctests.py::TestDoctests::test_docstring[Series.value_counts] - AssertionError: 1 of 11 doctests failed for Series.value_counts:
  **********************************************************************
  File "/opt/conda/envs/test/lib/python3.10/site-packages/cudf/core/series.py", line 212, in Series.value_counts
  Failed example:
      s.value_counts(bins=3)
  Expected:
      (2.0, 3.0]      2
      (0.996, 2.0]    2
      (3.0, 4.0]      1
      dtype: int64
  Got:
      (2.0, 2.0]        2
      (0.996, 0.996]    2
      (3.0, 3.0]        1
      dtype: int64
  **********************************************************************
  1 items had failures:
     1 of  11 in Series.value_counts
  ***Test Failed*** 1 failures.
  
assert not 1
 +  where 1 = TestResults(failed=1, attempted=11).failed
=== 14 failed, 101786 passed, 2061 skipped, 916 xfailed in 875.80s (0:14:35) ===
```

**Comment by mroeschke** [comment]
Thanks yeah it appears I have an off-by-one error somewhere I'm trying to track down

**Pull request merged**